utils/controller.js

```diff
@@ -539,30 +542,34 @@ class ABServiceController extends EventEmitter {
                   // Prevent Big JSON data structures from making us unresponsive:
                   // The underlying cote library uses generic JSON.stringify() to
                   // prepare the data for sending across the network.  This can cause
-                  // our service to hang and be unresponsive.  So we will first use
-                  // an async friendly bfj.stringify() to encode our data as a string
+                  // our service to hang and be unresponsive. So we will first use
+                  // an async friendly JSON.stringify() in workerpool lib to encode our data as a string
                   // and then send it off to cote.
                   // This will be slightly slower, but be non blocking for the
                   // incoming requests and as a result will not cause us to crash
                   // or loose subsequent requests.
-                  abReq.performance.mark("bfj.stringify", { op: "serialize" });
+                  abReq.performance.mark("worker.JSON.stringify", {
+                     op: "serialize",
+                  });
                   abReq.emit("status", "stringifying response");
-                  bfj.stringify(data, {
-                     /* options */
-                  })
-                     .then((strResponse) => {
-                        abReq.performance.measure("bfj.stringify");
-                        abReq.performance.log();
-                        abReq.spanEnd(handler.key);
-                        endRequest(abReq.requestID, cbErr, strResponse);
-                     })
-                     .catch((error) => {
-                        // :(
-                        abReq.log("ERROR bfj.stringify()", error);
-                        abReq.performance.log();
-
-                        endRequest(abReq.requestID, cbErr, data);
-                     });
+
+                  try {
+                     const strResponse = await this.worker(
+                        (json) => JSON.stringify(json),
+                        [data],
+                     );
+
+                     abReq.performance.measure("worker.JSON.stringify");
+                     abReq.performance.log();
+                     abReq.spanEnd(handler.key);
+                     endRequest(abReq.requestID, cbErr, strResponse);
+                  } catch (error) {
+                     // :(
+                     abReq.log("ERROR worker.JSON.stringify()", error);
```

Comment:
So I attempted to stringify and compare both (workerpool and bfj) in a 1000-length array of the kitchensink app. I encountered this error for bfj.

```
<--- Last few GCs --->

[47835:0x148078000]    81153 ms: Scavenge 2038.8 (2073.0) -> 2035.1 (2073.0) MB, 2.2 / 0.0 ms  (average mu = 0.266, current mu = 0.241) allocation failure; 
[47835:0x148078000]    81157 ms: Scavenge 2038.9 (2073.0) -> 2035.2 (2073.0) MB, 2.0 / 0.0 ms  (average mu = 0.266, current mu = 0.241) allocation failure; 
[47835:0x148078000]    81161 ms: Scavenge 2039.0 (2073.0) -> 2035.3 (2081.0) MB, 2.1 / 0.0 ms  (average mu = 0.266, current mu = 0.241) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1027094fc node::Abort() [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 2: 0x1027096ec node::ModifyCodeGenerationFromStrings(v8::Local<v8::Context>, v8::Local<v8::Value>, bool) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 3: 0x10285f0fc v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 4: 0x102a0a088 v8::internal::EmbedderStackStateScope::EmbedderStackStateScope(v8::internal::Heap*, v8::internal::EmbedderStackStateScope::Origin, cppgc::EmbedderStackState) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 5: 0x102a0dc90 v8::internal::Heap::CollectSharedGarbage(v8::internal::GarbageCollectionReason) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 6: 0x102a0ac84 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*, v8::GCCallbackFlags) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 7: 0x102a080dc v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 8: 0x102a07204 v8::internal::Heap::HandleGCRequest() [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
 9: 0x1029b409c v8::internal::StackGuard::HandleInterrupts() [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
10: 0x102d72d6c v8::internal::Runtime_StackGuardWithGap(int, unsigned long*, v8::internal::Isolate*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
11: 0x1030bd04c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
12: 0x107eb27d8 
13: 0x10308d178 Builtins_LoadGlobalIC [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
14: 0x107e498bc 
15: 0x107eb4b90 
16: 0x107ed3a08 
17: 0x107eadd50 
18: 0x107eaa6e8 
19: 0x107eb11f0 
20: 0x107e27714 
21: 0x107e4dbdc 
22: 0x107ecddfc 
23: 0x107ecdf98 
24: 0x107ed25cc 
25: 0x1030464d0 Builtins_JSEntryTrampoline [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
26: 0x103046164 Builtins_JSEntry [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
27: 0x10298b85c v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
28: 0x10298ad90 v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
29: 0x10287b124 v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
30: 0x102658eec node::InternalMakeCallback(node::Environment*, v8::Local<v8::Object>, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
31: 0x1026592c4 node::MakeCallback(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*, node::async_context) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
32: 0x1026b3618 node::Environment::CheckImmediate(uv_check_s*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
33: 0x10302f358 uv__run_check [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
34: 0x103029020 uv_run [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
35: 0x1026596f4 node::SpinEventLoop(node::Environment*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
36: 0x1027468ec node::NodeMainInstance::Run() [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
37: 0x1026d6028 node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult const*) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
38: 0x1026d62b4 node::Start(int, char**) [/Users/nohrsom/.nvm/versions/node/v18.16.0/bin/node]
39: 0x189d09058 start [/usr/lib/dyld]
zsh: abort      npm start
```

Then I reduced the length to 500 and ran the test. This is the benchmark.
<img width="642" alt="Screenshot 2566-11-10 at 15 23 49" src="https://github.com/digi-serve/ab-utils/assets/80669791/c1c4f1e2-7bd0-4c7a-93e5-fcf3deaa898e">

Reply:
When sending data to a worker pool, nodejs now uses `Transferable Objects` ( https://developer.chrome.com/blog/transferable-objects-lightning-fast/ ).  

However, when sending a variable from the main thread to the worker thread, that variable becomes unusable in the main thread after that.  ( https://nodejs.org/dist/latest-v20.x/docs/api/worker_threads.html#considerations-when-transferring-typedarrays-and-buffers )

So most likely the problem has to do with sending the `dataJSON` to the worker pool before you send it to `BFJ`.  

Try to send it to `BFJ` first, then send it to the `Workerpool`.  

If that fixes things, that was the problem.  If you need to test the Workerpool first, then clone the `dataJSON` and send a cloned copy to `BFJ`

Reply:
Ok so I did research initially. And `workerpool` is using node `worker_threads` under the hood. This uses the `structuredClone` algorithm to send data to the worker. I was also unsure how it would perform vs `JSON.stringify()`. When I asked @guyyoo  to benchmark, I didn't consider the blocking time, but just the overall time. I through together a quick test. And I'm now not sure `that workerpool` is a good solution here.

Generally the blocking time for using the worker thread is only a bit less than JSON.stringify()
<details><summary>Blocking Time/Total Time measurements</summary>
<p>

----------- 0 x kitchensink -----------
bfj - blocking time: 35.9012 ms
bfj - total time:    43.1269 ms
workerpool - blocking time: 4.5453 ms
workerpool - total time:    81.8987 ms
json - blocking time: 0.0848 ms
json - total time:    0.5659 ms
----------- 1 x kitchensink -----------
bfj - blocking time: 0.3737 ms
bfj - total time:    2029.74 ms
workerpool - blocking time: 31.4129 ms
workerpool - total time:    166.944 ms
json - blocking time: 30.9309 ms
json - total time:    31.2276 ms
----------- 2 x kitchensink -----------
bfj - blocking time: 0.458 ms
bfj - total time:    3984.7283 ms
workerpool - blocking time: 63.917 ms
workerpool - total time:    309.31 ms
json - blocking time: 78.6634 ms
json - total time:    79.0019 ms
----------- 3 x kitchensink -----------
bfj - blocking time: 0.3645 ms
bfj - total time:    5964.3853 ms
workerpool - blocking time: 97.2675 ms
workerpool - total time:    482.1223 ms
json - blocking time: 119.8279 ms
json - total time:    120.0993 ms
----------- 4 x kitchensink -----------
bfj - blocking time: 0.2171 ms
bfj - total time:    8343.8863 ms
workerpool - blocking time: 131.0585 ms
workerpool - total time:    685.9919 ms
json - blocking time: 161.1177 ms
json - total time:    161.3871 ms
----------- 5 x kitchensink -----------
bfj - blocking time: 0.5828 ms
bfj - total time:    10041.2772 ms
workerpool - blocking time: 160.8904 ms
workerpool - total time:    841.5534 ms
json - blocking time: 186.9855 ms
json - total time:    187.3894 ms
----------- 6 x kitchensink -----------
bfj - blocking time: 0.3033 ms
bfj - total time:    12002.7076 ms
workerpool - blocking time: 196.6308 ms
workerpool - total time:    1033.2279 ms
json - blocking time: 214.7628 ms
json - total time:    215.0528 ms
----------- 7 x kitchensink -----------
bfj - blocking time: 0.2819 ms
bfj - total time:    14364.9752 ms
workerpool - blocking time: 221.7682 ms
workerpool - total time:    1200.3301 ms
json - blocking time: 254.6125 ms
json - total time:    254.9204 ms
----------- 8 x kitchensink -----------
bfj - blocking time: 1.5058 ms
bfj - total time:    16136.7689 ms
workerpool - blocking time: 265.401 ms
workerpool - total time:    1416.1965 ms
json - blocking time: 302.9393 ms
json - total time:    303.5647 ms
----------- 9 x kitchensink -----------
bfj - blocking time: 0.248 ms
bfj - total time:    18132.7327 ms
workerpool - blocking time: 334.193 ms
workerpool - total time:    1718.3207 ms
json - blocking time: 386.1687 ms
json - total time:    386.4582 ms

</p>
</details> 

<details><summary>Code</summary>
<pre>
  <code>
function blockingTime(fn, name) {
  return new Promise(resolve => {
    var time = process.hrtime();
    let promise = fn();
    process.nextTick(async function() {
      var diff = process.hrtime(time);
      console.log('%s - blocking time: %d ms', name, (diff[0] * 1e9 + diff[1]) / 1e6);
      await promise;
      diff = process.hrtime(time);
      console.log('%s - total time:    %d ms', name, (diff[0] * 1e9 + diff[1]) / 1e6);
      resolve();
    });
  });
}
const tests = async () => {
  await blockingTime(bfjstr, "bfj");
  await blockingTime(worker, "workerpool");
  await blockingTime(json, "json");
}
const runTests = async () => {
while (payload.length < 10) {
  log(`----------- ${payload.length} x kitchensink -----------`);
  await run();
  payload.push(structuredClone((appJson)));
}}
runTests()
  </code>
</pre>
</details>

Reply:
I have even conducted testing with just await bfj.stringify() alone, but I encountered the same error. @Hiro-Nakamura

Reply:
https://github.com/digi-serve/ab_service_definition_manager/pull/64#discussion_r1393303775